### ISSUE TITLE
Does prometheus have a thread leak other CPU usage bug?

### ISSUE THREAD

**Issue by karlroberts** [prometheus-operator/kube-prometheus#747]
**What happened?**
Installed Prometheus Operator via a helm chart v 9.3.1 . after a few days we get CPUThrottling alerts. 
looking at the Kubernetes / Compute Resources / Pod dashboard in grafana over 7 days I can see a gradual rise in CPU until I start throttling see image attached 
![Screenshot from 2020-10-30 09-02-49](https://user-images.githubusercontent.com/462160/97639029-e5a04c80-1a91-11eb-9a85-c8c7fccaaf12.png)

**Did you expect to see some different?**
expected no alerts. we are not adding more work to the system or more exports and servicemonitors

**How to reproduce it (as minimally and precisely as possible)**:
deploy prometheus-operator to an environment and watch it for a few days

**Environment**
Production

* Prometheus Operator version:

    `quay.io/coreos/prometheus-operator:v0.38.1`
    
```
kubectl -n monitoring describe deployment prometheus-prometheus-oper-operator
Name:                   prometheus-prometheus-oper-operator
Namespace:              monitoring
CreationTimestamp:      Tue, 29 Sep 2020 08:49:40 +1000
Labels:                 app=prometheus-operator-operator
                        chart=prometheus-operator-9.3.1
                        heritage=Tiller
                        release=prometheus
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               app=prometheus-operator-operator,release=prometheus
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app=prometheus-operator-operator
                    chart=prometheus-operator-9.3.1
                    heritage=Tiller
                    release=prometheus
  Service Account:  prometheus-prometheus-oper-operator
  Containers:
   prometheus-operator:
    Image:      quay.io/coreos/prometheus-operator:v0.38.1
    Port:       8080/TCP
    Host Port:  0/TCP
    Args:
      --manage-crds=true
      --kubelet-service=kube-system/prometheus-prometheus-oper-kubelet
      --logtostderr=true
      --localhost=127.0.0.1
      --prometheus-config-reloader=quay.io/coreos/prometheus-config-reloader:v0.38.1
      --config-reloader-image=docker.io/jimmidyson/configmap-reload:v0.3.0
      --config-reloader-cpu=0
      --config-reloader-memory=0
    Environment:  <none>
    Mounts:       <none>
   tls-proxy:
    Image:      squareup/ghostunnel:v1.5.2
    Port:       8443/TCP
    Host Port:  0/TCP
    Args:
      server
      --listen=:8443
      --target=127.0.0.1:8080
      --key=cert/key
      --cert=cert/cert
      --disable-authentication
    Environment:  <none>
    Mounts:
      /cert from tls-proxy-secret (ro)
  Volumes:
   tls-proxy-secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  prometheus-prometheus-oper-admission
    Optional:    false
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   prometheus-prometheus-oper-operator-6dd4fd7466 (1/1 replicas created)
Events:          <none>

```

* Kubernetes version information:

    `Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-20T14:12:21Z", GoVersion:"go1.15.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.9-eks-4c6976", GitCommit:"4c6976793196d70bc5cd29d56ce5440c9473648e", GitTreeState:"clean", BuildDate:"2020-07-17T18:46:04Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}`
    

* Kubernetes cluster kind:
AWS EKS v 1.17

    insert how you created your cluster: kops, bootkube, tectonic-installer, etc.
terraform using this module [terraform-aws-eks](https://github.com/terraform-aws-modules/terraform-aws-eks)

* Manifests:

```
insert manifests relevant to the issue
```

* Prometheus Operator Logs:

```
Insert Prometheus Operator logs relevant to the issue here
```

* Prometheus Logs:

```
Insert Prometheus logs relevant to the issue here
```

**Anything else we need to know?**:
I have not added any rules other than the defaults that come with prometheus-operator helm chart. I see the same behaviour in an brand new EKS kubernetes environment with nothing deployed except propmetheus-operator and nginx-ingress and external-dns

**Comment by brancz** [comment]
I realize this may be a big ask, but do you think you could collect a couple of cpu profiles and share them? Ideally at different points where the CPU usage is low and high.

**Comment by karlroberts** [comment]
Hi Sorry I missed your question, ok. but what do you mean specifically?  CPU profile of the Pod? or the underlying host?
It's running in AWS EKS on their "optomised" linux AMI, do you know what command you want me to run there? 

I'll get a profile when I restart the POD when it always is low and another 9 days later.

**Comment by brancz** [comment]
My apologies, I should have elaborated a bit more. Prometheus has an HTTP endpoint that allows taking CPU profiles. We also have a small debug command in `promtool`. You can grab promtool by downloading a [latest Prometheus release](https://github.com/prometheus/prometheus/releases/tag/v2.23.0), and unpack the tarball. Then port-forward your Prometheus server:

```
kubectl -n monitoring port-forward prometheus-k8s-0 9090
```

And run

```
promtool debug all http://localhost:9090/
```

And then share the result here.

**Comment by karlroberts** [comment]
With that tool it wasn't a big ask at all :-) here are the two debug traces one busy and throttling after 8days of running the other 5 mins after I bounced the prometheus pod. Hope it helps :-) 
[not_busy_cpu_restart_5mins_debug.tar.gz](https://github.com/prometheus-operator/kube-prometheus/files/5639420/not_busy_cpu_restart_5mins_debug.tar.gz)
[busy_cpu_8days_debug.tar.gz](https://github.com/prometheus-operator/kube-prometheus/files/5639421/busy_cpu_8days_debug.tar.gz)

**Comment by brancz** [comment]
I had a quick glance at it appears to me that the "busier" server has a lot more churn of series going active and stale.

```
brancz at localhost in ~/Downloads/busy_cpu_8days_debug
$ cat metrics.txt | grep prometheus_tsdb_head_series
# HELP prometheus_tsdb_head_series Total number of series in the head block.
# TYPE prometheus_tsdb_head_series gauge
prometheus_tsdb_head_series 68849
# HELP prometheus_tsdb_head_series_created_total Total number of series created in the head
# TYPE prometheus_tsdb_head_series_created_total counter
prometheus_tsdb_head_series_created_total 132071
# HELP prometheus_tsdb_head_series_not_found_total Total number of requests for series that were not found.
# TYPE prometheus_tsdb_head_series_not_found_total counter
prometheus_tsdb_head_series_not_found_total 0
# HELP prometheus_tsdb_head_series_removed_total Total number of series removed in the head
# TYPE prometheus_tsdb_head_series_removed_total counter
prometheus_tsdb_head_series_removed_total 63222

brancz at localhost in ~/Downloads/busy_cpu_8days_debug
$ cat ../not_busy_cpu_restart_5mins_debug/metrics.txt | grep prometheus_tsdb_head_series
# HELP prometheus_tsdb_head_series Total number of series in the head block.
# TYPE prometheus_tsdb_head_series gauge
prometheus_tsdb_head_series 68871
# HELP prometheus_tsdb_head_series_created_total Total number of series created in the head
# TYPE prometheus_tsdb_head_series_created_total counter
prometheus_tsdb_head_series_created_total 68871
# HELP prometheus_tsdb_head_series_not_found_total Total number of requests for series that were not found.
# TYPE prometheus_tsdb_head_series_not_found_total counter
prometheus_tsdb_head_series_not_found_total 0
# HELP prometheus_tsdb_head_series_removed_total Total number of series removed in the head
# TYPE prometheus_tsdb_head_series_removed_total counter
prometheus_tsdb_head_series_removed_total 0
```

Could you share graphs of `prometheus_tsdb_head_series`,  `prometheus_tsdb_head_series_created_total`, and `prometheus_tsdb_head_series_removed_total` over time? That should correlate to the CPU graphs.

**Comment by karlroberts** [comment]
Is this what you need? remember I bounced the prometheus pod to get the low cpu debug info.

prometheus_tsdb_head_series:

prometheus_tsdb_head_series{endpoint="web",instance="172.17.1.153:9090",job="prometheus-prometheus-oper-prometheus",namespace="monitoring",pod="prometheus-prometheus-prometheus-oper-prometheus-0",service="prometheus-prometheus-oper-prometheus"} | 60900
![Screenshot from 2020-12-05 09-22-44](https://user-images.githubusercontent.com/462160/101221304-890cfe80-36db-11eb-958d-51bc4c32cde4.png)



prometheus_tsdb_head_series_created_total: 

prometheus_tsdb_head_series_created_total{endpoint="web",instance="172.17.1.153:9090",job="prometheus-prometheus-oper-prometheus",namespace="monitoring",pod="prometheus-prometheus-prometheus-oper-prometheus-0",service="prometheus-prometheus-oper-prometheus"} | 110206
![Screenshot from 2020-12-05 09-24-03](https://user-images.githubusercontent.com/462160/101221389-b8237000-36db-11eb-8a8e-08458efb993b.png)



prometheus_tsdb_head_series_removed_total
prometheus_tsdb_head_series_removed_total{endpoint="web",instance="172.17.1.153:9090",job="prometheus-prometheus-oper-prometheus",namespace="monitoring",pod="prometheus-prometheus-prometheus-oper-prometheus-0",service="prometheus-prometheus-oper-prometheus"} | 49306
![Screenshot from 2020-12-05 09-26-16](https://user-images.githubusercontent.com/462160/101221524-f882ee00-36db-11eb-8d09-62f755f6f457.png)

**Comment by brancz** [comment]
I'm going to guess the first graph(s) are from the high CPU Prometheus server?

**Comment by simonpasquier** [comment]
Looking quickly at the CPU profile, it looks like most of the time is spent in rules evaluation. From the collected metrics, the group that takes the longest time to evaluate is `kube-apiserver.rules`. I suspect you're hitting something similar to #503.

**Comment by karlroberts** [comment]
ok so what do I do about it? I just have the rules as set by the Prometheus operator helm chart. do I need to override them or just wait for another release?

**Comment by karlroberts** [comment]
FYI that first graph was after the High CPU prometheus was restarted

**Comment by karlroberts** [comment]
FYI I stopped monitoring apiserver and CPU dropped nicely. Thanks guys. I'll Close this Issue for now, we will be migrating to a later version of prometheus-operator (now kube-prometheus-stack) after our Christmas code freeze. I'll re-instate the monitoring of api server then to see if the fixes mentioned in [#503](https://github.com/prometheus-operator/kube-prometheus/issues/503) worked for us too
